lib/internal/test_runner/yaml_parser.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+const {
+  codes: {
+    ERR_TEST_FAILURE,
+  }
+} = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const {
+  ArrayPrototypeJoin,
+  Error,
+  Number,
+  NumberIsNaN,
+  RegExpPrototypeExec,
+  StringPrototypeEndsWith,
+  StringPrototypeRepeat,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+  StringPrototypeSubstring,
+} = primordials;
+
+const kYamlKeyRegex = /^(\s+)?(\w+):(\s)+([>|][-+])?(.*)$/;
+const kStackDelimiter = '    at ';
+
+function reConstructError(parsedYaml) {
+  if (!('error' in parsedYaml)) {
+    return parsedYaml;
+  }
+  const isAssertionError = parsedYaml.code === 'ERR_ASSERTION' ||
+    'actual' in parsedYaml || 'expected' in parsedYaml || 'operator' in parsedYaml;
+  const isTestFauilure = parsedYaml.code === 'ERR_TEST_FAILURE' || 'failureType' in parsedYaml;
+  const stack = parsedYaml.stack ? kStackDelimiter + ArrayPrototypeJoin(parsedYaml.stack, `\n${kStackDelimiter}`) : '';
+
+  // eslint-disable-next-line no-restricted-syntax
+  let error = new Error(parsedYaml.error);
+  error.stack = stack;
+  error.code = parsedYaml.code;
```

Comment:
Should this be an `else` condition on the `if (isAssertionError)` below?

EDIT: Nevermind, it looks like that is what Antoine is suggesting already.